src/front/spv.rs

```diff
@@ -2055,51 +2148,86 @@ impl<I: Iterator<Item = u32>> Parser<I> {
                 Instruction { op, .. } => return Err(Error::InvalidParameter(op)),
             }
         }
-        // read body
+
+        // Read body
         let mut local_function_calls = FastHashMap::default();
-        let mut control_flow_graph = FastHashMap::default();
+        let mut control_flow_graph = Dag::<ControlFlowNode, ControlFlowEdge, u32>::new();
+        let mut id_to_node = std::collections::HashMap::<u32, NodeIndex<u32>>::new();
+
+        // Scan the blocks and add them as nodes
         loop {
             let fun_inst = self.next_inst()?;
             log::debug!("\t\t{:?}", fun_inst.op);
             match fun_inst.op {
                 spirv::Op::Label => {
+                    // Read the label ID
                     fun_inst.expect(2)?;
                     let label_id = self.next()?;
+
                     let node = self.next_block(
+                        label_id,
                         &mut fun.expressions,
                         &mut fun.local_variables,
                         &module.types,
                         &module.constants,
                         &module.global_variables,
                         &mut local_function_calls,
                     )?;
-                    // temp until the CFG is fully processed
-                    for assign in node.assignments.iter() {
-                        fun.body.push(crate::Statement::Store {
-                            pointer: assign.to,
-                            value: assign.value,
-                        });
-                    }
-                    match node.terminator {
-                        Terminator::Return { value } => {
-                            fun.body.push(crate::Statement::Return { value });
-                        }
-                        Terminator::Branch {
-                            label_id,
-                            condition,
-                        } => {
-                            let _ = (label_id, condition); //TODO
-                        }
-                    }
-                    control_flow_graph.insert(label_id, node);
+
+                    id_to_node.insert(label_id, control_flow_graph.add_node(node));
                 }
                 spirv::Op::FunctionEnd => {
                     fun_inst.expect(1)?;
                     break;
                 }
-                _ => return Err(Error::UnsupportedInstruction(self.state, fun_inst.op)),
+                _ => panic!("SHOULD NOT HAPPEN"),
+            }
+        }
+
+        // Create Edges.
+        let mut edges = Vec::new();
+        for node in control_flow_graph.node_weights_mut() {
```

Comment:
is it feasible to try adding the edges in the same place where we add the nodes, so that we avoid this phase completely?

Reply:
I was thinking about It but I think even a simple example shows that no. See: http://shader-playground.timjones.io/4e94c99f421a7382cb4236456890d55d. OpSwitch points to block that have not been declared yet by the time parser arrived at the instruction.

I also messed up a little and will have to switch from daggy to pure graph. The graph can actually have cycles and what not. So we really have to first build nodes, then edges and after that you can finish the whole classification.